pkg/generator/appconfiguration/generator/monitoring_generator.go

```diff
@@ -0,0 +1,121 @@
+package generator
+
+import (
+	"fmt"
+
+	prometheusV1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"kusionstack.io/kusion/pkg/generator/appconfiguration"
+	"kusionstack.io/kusion/pkg/models"
+	"kusionstack.io/kusion/pkg/models/appconfiguration/monitoring"
+	"kusionstack.io/kusion/pkg/projectstack"
+)
+
+type monitoringGenerator struct {
+	project *projectstack.Project
+	monitor *monitoring.Monitor
+	appName string
+}
+
+func NewMonitoringGenerator(project *projectstack.Project, monitor *monitoring.Monitor, appName string) (appconfiguration.Generator, error) {
+	if len(project.Name) == 0 {
+		return nil, fmt.Errorf("project name must not be empty")
+	}
+
+	if len(appName) == 0 {
+		return nil, fmt.Errorf("app name must not be empty")
+	}
+	return &monitoringGenerator{
+		project: project,
+		monitor: monitor,
+		appName: appName,
+	}, nil
+}
+
+func NewMonitoringGeneratorFunc(project *projectstack.Project, monitor *monitoring.Monitor, appName string) appconfiguration.NewGeneratorFunc {
+	return func() (appconfiguration.Generator, error) {
+		return NewMonitoringGenerator(project, monitor, appName)
+	}
+}
+
+func (g *monitoringGenerator) Generate(spec *models.Spec) error {
+	if spec.Resources == nil {
+		spec.Resources = make(models.Resources, 0)
+	}
+
+	monitoringLabels := map[string]string{
+		"kusion_monitoring_appname": g.appName,
+	}
+
+	if g.project.ProjectConfiguration.Prometheus != nil && g.project.ProjectConfiguration.Prometheus.OperatorMode {
```

Comment:
Is it possible for `ProjectConfiguration` to be nil in this context?

Reply:
I don't think so. The `config` object gets [initialized in `ParseProjectConfiguration`](https://github.com/KusionStack/kusion/blob/main/pkg/projectstack/project.go#L68-L75) and then its address gets [passed around into the `project` as project.ProjectConfiguration](https://github.com/KusionStack/kusion/blob/main/pkg/projectstack/project.go#L100-L120).

The only scenario there _might_ be a problem is if project.yaml is empty. But in that case it won't be able to render anything because it's missing the generator type and project name.